### ISSUE TITLE
Is shared bundle generation possible with static builds?

### ISSUE THREAD

**Issue by pramodsankarl** [systemjs/builder#564]
I started using the systemjs builder for my external modules (on js files transpiled by TypeScript in SystemJs mode). Since i need to use self executing bundle i started using `buildStatic` and works awesome. However i am having an issue while determining the common trace and creating shared bundles using trace api as well as using arithmetic operation.

For example:

         Promise.all([builder.trace('components/person/person.js'), builder.trace('services/services.js')])
            .then(function(trees) {
                var commonTree = builder.intersectTrees(trees[0], trees[1]);
                return Promise.all([
                    builder.buildStatic(commonTree, 'app/shared-bundle.js', {minify: true, mangle: true, sourcemaps:false}),
                    builder.buildStatic(builder.subtractTrees(trees[0], commonTree), 'app/person.min.js', {minify: true, mangle: true, sourcemaps:false}),
                    builder.buildStatic(builder.subtractTrees(trees[1], commonTree), 'app/services.js', {minify: true, mangle: true, sourcemaps:false})
                ]);
            }).then....

In the above example it generates common tree however while passing it into `buildStatic` it fails in [this line](https://github.com/systemjs/builder/blob/master/lib/builder.js#L616). The common tree generates is an Object with properties and not an array so effectively `entryPoints` is undefined there after failing to enter the previous conditions.

![image](https://cloud.githubusercontent.com/assets/3045484/14660592/b68b1dde-0674-11e6-9fc2-ccd866778a9e.png)

So just wanted to cross check if this is as expected or i might be missing something. This is a very simple example at this point. 

Appreciate your help.

**Comment by gcnm-pgeil** [comment]
This is not possible for static builds since each one has everything it needs "hard wired" into the final bundle code. Even module identifiers are replaced during compilation with arbitrary incrementing numbers asf., they contain no loader config or any other kind of dependency informations. Think of static bundles as kind of a blackbox with only certain public properties -- nothing more. They are not meant for being used in a dynamic systemjs module loading environment. Use the default `bundle`  method instead which simply bundles "consumeable" modules together.

**Comment by pramodsankarl** [comment]
Thanks @gcnm-pgeil. Say I need creating multiple bundles and load a subset of them, meanwhile i do am trying to avoid any shared dependencies from being duplicated in those.  

In the documentation under "Bundle Arithmetic" it is mentioned that `buildStatic` also supports arithmetic operations as well as passing in the build trace. So for example. isn't `builder.intersectTrees(trees[0], trees[1]);` (from my above example) equivalent to `.buildStatic('components/person/person.js & services/services.js','shared.js',...` and similarly `buildStatic(builder.subtractTrees(trees[0], commonTree),...` equivalent to `.buildStatic('components/person/person.js - shared.js'` ?

**Comment by amenadiel** [comment]
I get you want to do this automatically and I don't know the answer. However, what I'm doing in a similar situation is having a common (`lib_common.js`) entry point that provides a reference to common libraries:

```js
define([
    'jquery',
    'underscore',
    'backbone',
    'd3',
    'bluebird'
],function(jQuery, _, Backbone, d3, Promise) {
    var CommonDeps: {
        jQuery : jQuery,
        _ : _,
        backbone : backbone,
        d3 : d3,
        Promise : Promise
    };
    return CommonDeps;
});
```

In every other static build I need to create, I do no longer request the libraries directly. Instead, I list `lib_common` as a dependency and use its references to the libraries.

So, instead of

```js
define([
    'jquery',
    'underscore',
    'backbone'
], function(jQuery, _, Backbone) {
    ...
});
```

I do

```js
define([
    'lib_common'
],function(CommonDeps) {
    var jQuery = CommonDeps.jQuery,
          _ = CommonDeps._,
          Backbone = CommonDeps.Backbone
    ...
});
```

On the build stage, I substract `lib_common.js` from the arithmetic build, and use the `globalDeps` option to map it to `lib_common` 

```js
builder.config({
    globalDeps : {
        "lib_common":"lib_common"
    }
});
...

So every SFX file I create ends with:

```js
(function(factory) {
  define(["lib_common"], factory);
});
```

**Comment by guybedford** [comment]
Fix added in https://github.com/systemjs/builder/commit/5245986b781db05d96710d1c03f4c4aab79c6c6d. Let me know if that helps.

**Comment by gcnm-pgeil** [comment]
I just want to clarify: The actual (functional) issue was that result objects of `buildStatic` could not be used for further tracing / bundle arithmetic operations. This seems to be addressed in the recent commits of guybedford. 

The other question -- which I tried to answer -- was whether it should be considered a good practice to use static bundle files from within other static bundles files by using an *adapter module* whose sole purpose is to provide importable references to the exported values of a "common bundle". Albeit this might work,I consider it an awkward solution to a problem for that systemjs offers a more elegant and efficient solution through its actual dependency-/loading functionality for dynamic modules.

So instead of manually writing adapter modules for "common bundles", I would  compile everything into dynamic module bundle files, throw in the systemjs runtime and get all the module loading and dependency resolval stuff for free out of the box. When static dead code elemination (rollup) is not a criteria, one could also argue that by doing so we also avoid all the duplicate code which whill otherwise be inevitably contained in every static bundle being generated (not the common libs, but common runtime boilerplate code asf.).

**Comment by pramodsankarl** [comment]
@guybedford That works great. Thanks a lot for the quick fix. 
@amenadiel Thanks a lot for the elaborated example. I wasn't sure how to load partial bundles, but i have a clue now from your examples.
@gcnm-pgeil Thanks a lot for your suggestion, in fact due to a specific scenario in my project (which is very huge, a mix of traditional scripts references, external modules and have to deal with specific user permission based loading of bundles dynamically in an angular 1.x app) i wanted to explore self executing bundle option. But i totally agree your suggestion. I do have to figure out how can i fit in dynamic loading with all the constraints i have in the application (i do that already in the Dev mode, however my release mode has a bit more complexity in the application).

**Comment by amenadiel** [comment]
@gcnm-pgeil I used to avoid static builds because I agree is convoluted, but It often happens that different bundles contain references to specific versions of its dependencies, and install deduping only goes so far, so you either end with 4 different forks, or an unmet dependency. 

I guess I'm just doing a wrong workflow, but in my eyes is much better to require/define from a sfx getting a grab of the imported module, instead of doing System.import from a bundle that doesn't return, and then hoping for the best before doing System.get

**Comment by gcnm-pgeil** [comment]
@amenadiel I see you point. I didn't have to deal with multiple version-specific deps to the same module/package so I have no experience to share here. What counts in the end is that we get things up and running.

**Comment by guybedford** [comment]
Released in 0.15.16.
